File: src/com/dafttech/terra/game/Events.java
package com.dafttech.terra.game;

import static com.dafttech.terra.resources.Options.BLOCK_SIZE;

import com.badlogic.gdx.Gdx;
import com.dafttech.eventmanager.Event;
import com.dafttech.eventmanager.EventManager;
import com.dafttech.eventmanager.EventType;
import com.dafttech.eventmanager.ListenerContainer;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.tiles.Tile;

public class Events {
    public static final EventManager EVENTMANAGER = new EventManager();

    public static final EventType EVENT_INITPRE = new EventType("INITPRE", EVENTMANAGER);
    public static final EventType EVENT_INITPOST = new EventType("INITPOST", EVENTMANAGER);

    public static final EventType EVENT_WINRESIZE = new EventType("WINRESIZE", EVENTMANAGER);
    public static final EventType EVENT_WINPAUSE = new EventType("WINPAUSE", EVENTMANAGER);
    public static final EventType EVENT_WINRESUME = new EventType("WINRESUME", EVENTMANAGER);
    public static final EventType EVENT_WINDISPOSE = new EventType("WINDISPOSE", EVENTMANAGER);

    public static final EventType EVENT_BLOCKCHANGE = new EventType("BLOCKCHANGE", EVENTMANAGER);

    public static final EventType EVENT_CHATCOMMAND = new EventType("CHATCOMMAND", EVENTMANAGER);

    public static final EventType EVENT_KEYDOWN = new EventType("KEYDOWN", EVENTMANAGER) {
        @Override
        protected boolean applyFilter(Event event, Object[] filter, ListenerContainer eventListener) {
            return ((String) filter[0]).equals(event.getInput(0, String.class));
        }
    };
    public static final EventType EVENT_KEYUP = new EventType("KEYUP", EVENTMANAGER) {
        @Override
        protected boolean applyFilter(Event event, Object[] filter, ListenerContainer eventListener) {
            return ((String) filter[0]).equals(event.getInput(0, String.class));
        }
    };

    public static final EventType EVENT_MOUSEDOWN = new EventType("MOUSEDOWN", EVENTMANAGER);
    public static final EventType EVENT_MOUSEUP = new EventType("MOUSEUP", EVENTMANAGER);
    public static final EventType EVENT_MOUSEMOVE = new EventType("MOUSEMOVE", EVENTMANAGER);
    public static final EventType EVENT_SCROLL = new EventType("SCROLL", EVENTMANAGER);

    public static final EventType EVENT_BLOCKUPDATE = new EventType("BLOCKUPDATE", EVENTMANAGER) {
        @Override
        protected void onEvent(Event event) {
            Tile tile = event.getInput(0, World.class).getTile(event.getInput(1, Integer.class), event.getInput(2, Integer.class));
            if (tile != null) tile.update(event.getInput(0, World.class), 0);
        }
    };

    public static final EventType EVENT_WORLDTICK = new EventType("GAMETICK", EVENTMANAGER) {
        @Override
        protected void onEvent(Event event) {
            World world = event.getInput(0, World.class);
            int sx = 25 + Gdx.graphics.getWidth() / BLOCK_SIZE / 2;
            int sy = 25 + Gdx.graphics.getHeight() / BLOCK_SIZE / 2;
            Tile tile;
            for (int x = (int) world.localPlayer.getPosition().x / BLOCK_SIZE - sx; x < (int) world.localPlayer.getPosition().x / BLOCK_SIZE + sx; x++) {
                for (int y = (int) world.localPlayer.getPosition().y / BLOCK_SIZE - sy; y < (int) world.localPlayer.getPosition().y / BLOCK_SIZE + sy; y++) {
                    tile = world.getTile(x, y);
                    if (tile != null) tile.tick(world, event.getInput(1, Float.class));
                }
            }
        }
    };

    public static void init() {
    }
}

File: src/main/scala/com/dafttech/terra/game/Events.scala
package com.dafttech.terra.game

import com.badlogic.gdx.Gdx
import com.dafttech.terra.game.world.World
import com.dafttech.terra.game.world.tiles.Tile
import com.dafttech.terra.resources.Options.BLOCK_SIZE
import org.lolhens.eventmanager.{Event, EventManager, EventType, ListenerContainer}
import org.lolhens.storage.tuple.Tuple

object Events {
  val EVENTMANAGER: EventManager = new EventManager
  val EVENT_INITPRE: EventType = new EventType("INITPRE", EVENTMANAGER)
  val EVENT_INITPOST: EventType = new EventType("INITPOST", EVENTMANAGER)
  val EVENT_WINRESIZE: EventType = new EventType("WINRESIZE", EVENTMANAGER)
  val EVENT_WINPAUSE: EventType = new EventType("WINPAUSE", EVENTMANAGER)
  val EVENT_WINRESUME: EventType = new EventType("WINRESUME", EVENTMANAGER)
  val EVENT_WINDISPOSE: EventType = new EventType("WINDISPOSE", EVENTMANAGER)
  val EVENT_BLOCKCHANGE: EventType = new EventType("BLOCKCHANGE", EVENTMANAGER)
  val EVENT_CHATCOMMAND: EventType = new EventType("CHATCOMMAND", EVENTMANAGER)
  val EVENT_KEYDOWN: EventType = new EventType("KEYDOWN", EVENTMANAGER) {
    protected override def isFiltered(event: Event, filter: Tuple, eventListener: ListenerContainer): Boolean = {
      return filter.get(0, classOf[String]) == event.in.get(0, classOf[String])
    }
  }
  val EVENT_KEYUP: EventType = new EventType("KEYUP", EVENTMANAGER) {
    protected override def isFiltered(event: Event, filter: Tuple, eventListener: ListenerContainer): Boolean = {
      return filter.get(0, classOf[String]) == event.in.get(0, classOf[String])
    }
  }
  val EVENT_MOUSEDOWN: EventType = new EventType("MOUSEDOWN", EVENTMANAGER)
  val EVENT_MOUSEUP: EventType = new EventType("MOUSEUP", EVENTMANAGER)
  val EVENT_MOUSEMOVE: EventType = new EventType("MOUSEMOVE", EVENTMANAGER)
  val EVENT_SCROLL: EventType = new EventType("SCROLL", EVENTMANAGER)
  val EVENT_BLOCKUPDATE: EventType = new EventType("BLOCKUPDATE", EVENTMANAGER) {
    protected override def onEvent(event: Event) {
      val tile: Tile = event.in.get(0, classOf[World]).getTile(event.in.get(1, classOf[Integer]), event.in.get(2, classOf[Integer]))
      if (tile != null) tile.update(event.in.get(0, classOf[World]), 0)
    }
  }
  val EVENT_WORLDTICK: EventType = new EventType("GAMETICK", EVENTMANAGER) {
    protected override def onEvent(event: Event) {
      val world: World = event.in.get(0, classOf[World])
      val sx: Int = 25 + Gdx.graphics.getWidth / BLOCK_SIZE / 2
      val sy: Int = 25 + Gdx.graphics.getHeight / BLOCK_SIZE / 2
      var tile: Tile = null
      var x: Int = world.localPlayer.getPosition.x.toInt / BLOCK_SIZE - sx
      while (x < world.localPlayer.getPosition.x.toInt / BLOCK_SIZE + sx) {
        {
          {
            var y: Int = world.localPlayer.getPosition.y.toInt / BLOCK_SIZE - sy
            while (y < world.localPlayer.getPosition.y.toInt / BLOCK_SIZE + sy) {
              {
                tile = world.getTile(x, y)
                if (tile != null) tile.tick(world, event.in.get(1, classOf[Float]))
              }
              ({
                y += 1;
                y - 1
              })
            }
          }
        }
        ({
          x += 1;
          x - 1
        })
      }
    }
  }

  def init {
  }
}